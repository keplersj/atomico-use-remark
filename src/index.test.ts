import { jest } from "@jest/globals";
import { render, h } from "atomico";
import { createHooks } from "atomico/test-hooks";
import { Plugin } from "unified";

afterEach(() => {
  document.querySelectorAll("html")[0].innerHTML = "";
});

describe("Remark Hook for Atomico", () => {
  describe("module", () => {
    it("matches expectations", async () => {
      const module_ = await import("./index");

      expect(module_).toMatchInlineSnapshot(`
        Object {
          "useRemark": [Function],
        }
      `);
    });
  });

  describe("hook, tree from initial content", () => {
    it("returns a result object when run", async () => {
      const host = document.createElement("div");
      host.attachShadow({ mode: "open" });
      document.body.append(host);
      const hooks = createHooks(jest.fn(), host);

      const { useRemark } = await import("./index");
      const remarkResult = hooks.load(() =>
        useRemark(`
# Test Content

*This* is **test content** for ***the*** \`remark\` hook for \`atomico\`.
      `)
      );

      expect(remarkResult).toMatchSnapshot();

      expect(document.documentElement).toMatchSnapshot();

      render(h("host", undefined, remarkResult[0]), host);

      expect(document.documentElement).toMatchSnapshot();
    });
  });

  describe("hook, tree from set content", () => {
    it("returns a result object when run", async () => {
      const host = document.createElement("div");
      host.attachShadow({ mode: "open" });
      document.body.append(host);
      const hooks = createHooks(() => {
        [tree, setContent] = hooks.load(() => useRemark());
        render(h("host", undefined, tree), host);
      }, host);

      const { useRemark } = await import("./index");
      let [tree, setContent] = hooks.load(() => useRemark());

      expect(tree).toMatchSnapshot();

      expect(document.documentElement).toMatchSnapshot();

      render(h("host", undefined, tree), host);

      expect(document.documentElement).toMatchSnapshot();

      setContent(`
# Set Test Content

*This* is **set test content** for ***the*** \`remark\` hook for \`atomico\`.
        `);

      expect(tree).toMatchSnapshot();

      expect(document.documentElement).toMatchSnapshot();
    });
  });

  describe("hook options", () => {
    it("accepts Remark Plugins", async () => {
      const host = document.createElement("div");
      host.attachShadow({ mode: "open" });
      document.body.append(host);
      const hooks = createHooks(jest.fn(), host);

      const mockRemarkPlugin = jest.fn();

      expect(mockRemarkPlugin.mock).toMatchInlineSnapshot(`
        Object {
          "calls": Array [],
          "instances": Array [],
          "invocationCallOrder": Array [],
          "results": Array [],
        }
      `);

      const { useRemark } = await import("./index");
      hooks.load(() =>
        useRemark("", { remarkPlugins: [mockRemarkPlugin as Plugin] })
      );

      expect(mockRemarkPlugin.mock).toMatchInlineSnapshot(`
        Object {
          "calls": Array [
            Array [],
          ],
          "instances": Array [
            [Function],
          ],
          "invocationCallOrder": Array [
            1,
          ],
          "lastCall": Array [],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      `);
    });

    it("accepts Rehype Plugins", async () => {
      const host = document.createElement("div");
      host.attachShadow({ mode: "open" });
      document.body.append(host);
      const hooks = createHooks(jest.fn(), host);

      const mockRehypePlugin = jest.fn();

      expect(mockRehypePlugin.mock).toMatchInlineSnapshot(`
        Object {
          "calls": Array [],
          "instances": Array [],
          "invocationCallOrder": Array [],
          "results": Array [],
        }
      `);

      const { useRemark } = await import("./index");
      hooks.load(() =>
        useRemark("", { rehypePlugins: [mockRehypePlugin as Plugin] })
      );

      expect(mockRehypePlugin.mock).toMatchInlineSnapshot(`
        Object {
          "calls": Array [
            Array [],
          ],
          "instances": Array [
            [Function],
          ],
          "invocationCallOrder": Array [
            2,
          ],
          "lastCall": Array [],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      `);
    });

    it("accepts Remark and Rehype Plugins", async () => {
      const host = document.createElement("div");
      host.attachShadow({ mode: "open" });
      document.body.append(host);
      const hooks = createHooks(jest.fn(), host);

      const mockRemarkPlugin = jest.fn();
      const mockRehypePlugin = jest.fn();

      expect(mockRemarkPlugin.mock).toMatchInlineSnapshot(`
        Object {
          "calls": Array [],
          "instances": Array [],
          "invocationCallOrder": Array [],
          "results": Array [],
        }
      `);

      expect(mockRehypePlugin.mock).toMatchInlineSnapshot(`
        Object {
          "calls": Array [],
          "instances": Array [],
          "invocationCallOrder": Array [],
          "results": Array [],
        }
      `);

      const { useRemark } = await import("./index");
      hooks.load(() =>
        useRemark("", {
          remarkPlugins: [mockRemarkPlugin as Plugin],
          rehypePlugins: [mockRehypePlugin as Plugin],
        })
      );

      expect(mockRemarkPlugin.mock).toMatchInlineSnapshot(`
        Object {
          "calls": Array [
            Array [],
          ],
          "instances": Array [
            [Function],
          ],
          "invocationCallOrder": Array [
            3,
          ],
          "lastCall": Array [],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      `);

      expect(mockRehypePlugin.mock).toMatchInlineSnapshot(`
        Object {
          "calls": Array [
            Array [],
          ],
          "instances": Array [
            [Function],
          ],
          "invocationCallOrder": Array [
            4,
          ],
          "lastCall": Array [],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      `);
    });
  });
});
