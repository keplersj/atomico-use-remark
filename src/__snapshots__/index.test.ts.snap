// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remark Hook for Atomico hook, tree from initial content returns a result object when run 1`] = `
Array [
  Object {
    "$$": Symbol(Atomico.$$),
    "children": Array [
      Array [
        Object {
          "$$": Symbol(Atomico.$$),
          "children": Array [
            Array [
              "Test Content",
            ],
          ],
          "clone": undefined,
          "is": undefined,
          "key": undefined,
          "props": Object {},
          "raw": false,
          "shadow": undefined,
          "static": undefined,
          "type": "h1",
        },
        "
",
        Object {
          "$$": Symbol(Atomico.$$),
          "children": Array [
            Array [
              Object {
                "$$": Symbol(Atomico.$$),
                "children": Array [
                  Array [
                    "This",
                  ],
                ],
                "clone": undefined,
                "is": undefined,
                "key": undefined,
                "props": Object {},
                "raw": false,
                "shadow": undefined,
                "static": undefined,
                "type": "em",
              },
              " is ",
              Object {
                "$$": Symbol(Atomico.$$),
                "children": Array [
                  Array [
                    "test content",
                  ],
                ],
                "clone": undefined,
                "is": undefined,
                "key": undefined,
                "props": Object {},
                "raw": false,
                "shadow": undefined,
                "static": undefined,
                "type": "strong",
              },
              " for ",
              Object {
                "$$": Symbol(Atomico.$$),
                "children": Array [
                  Array [
                    Object {
                      "$$": Symbol(Atomico.$$),
                      "children": Array [
                        Array [
                          "the",
                        ],
                      ],
                      "clone": undefined,
                      "is": undefined,
                      "key": undefined,
                      "props": Object {},
                      "raw": false,
                      "shadow": undefined,
                      "static": undefined,
                      "type": "strong",
                    },
                  ],
                ],
                "clone": undefined,
                "is": undefined,
                "key": undefined,
                "props": Object {},
                "raw": false,
                "shadow": undefined,
                "static": undefined,
                "type": "em",
              },
              " ",
              Object {
                "$$": Symbol(Atomico.$$),
                "children": Array [
                  Array [
                    "remark",
                  ],
                ],
                "clone": undefined,
                "is": undefined,
                "key": undefined,
                "props": Object {},
                "raw": false,
                "shadow": undefined,
                "static": undefined,
                "type": "code",
              },
              " hook for ",
              Object {
                "$$": Symbol(Atomico.$$),
                "children": Array [
                  Array [
                    "atomico",
                  ],
                ],
                "clone": undefined,
                "is": undefined,
                "key": undefined,
                "props": Object {},
                "raw": false,
                "shadow": undefined,
                "static": undefined,
                "type": "code",
              },
              ".",
            ],
          ],
          "clone": undefined,
          "is": undefined,
          "key": undefined,
          "props": Object {},
          "raw": false,
          "shadow": undefined,
          "static": undefined,
          "type": "p",
        },
      ],
    ],
    "clone": undefined,
    "is": undefined,
    "key": undefined,
    "props": Object {},
    "raw": false,
    "shadow": undefined,
    "static": undefined,
    "type": "div",
  },
  [Function],
]
`;

exports[`Remark Hook for Atomico hook, tree from initial content returns a result object when run 2`] = `
<html>
  <head />
  <body>
    <div />
  </body>
</html>
`;

exports[`Remark Hook for Atomico hook, tree from initial content returns a result object when run 3`] = `
<html>
  <head />
  <body>
    <div>
      Mark {}
      <div>
        Mark {}
        <h1>
          Mark {}
          Test Content
          Mark {}
        </h1>
        

        <p>
          Mark {}
          <em>
            Mark {}
            This
            Mark {}
          </em>
           is 
          <strong>
            Mark {}
            test content
            Mark {}
          </strong>
           for 
          <em>
            Mark {}
            <strong>
              Mark {}
              the
              Mark {}
            </strong>
            Mark {}
          </em>
           
          <code>
            Mark {}
            remark
            Mark {}
          </code>
           hook for 
          <code>
            Mark {}
            atomico
            Mark {}
          </code>
          .
          Mark {}
        </p>
        Mark {}
      </div>
      Mark {}
    </div>
  </body>
</html>
`;

exports[`Remark Hook for Atomico hook, tree from set content returns a result object when run 1`] = `
Object {
  "$$": Symbol(Atomico.$$),
  "children": Array [],
  "clone": undefined,
  "is": undefined,
  "key": undefined,
  "props": Object {},
  "raw": false,
  "shadow": undefined,
  "static": undefined,
  "type": "div",
}
`;

exports[`Remark Hook for Atomico hook, tree from set content returns a result object when run 2`] = `
<html>
  <head />
  <body>
    <div />
  </body>
</html>
`;

exports[`Remark Hook for Atomico hook, tree from set content returns a result object when run 3`] = `
<html>
  <head />
  <body>
    <div>
      Mark {}
      <div />
      Mark {}
    </div>
  </body>
</html>
`;

exports[`Remark Hook for Atomico hook, tree from set content returns a result object when run 4`] = `
Object {
  "$$": Symbol(Atomico.$$),
  "children": Array [
    Array [
      Object {
        "$$": Symbol(Atomico.$$),
        "children": Array [
          Array [
            "Set Test Content",
          ],
        ],
        "clone": undefined,
        "is": undefined,
        "key": undefined,
        "props": Object {},
        "raw": false,
        "shadow": undefined,
        "static": undefined,
        "type": "h1",
      },
      "
",
      Object {
        "$$": Symbol(Atomico.$$),
        "children": Array [
          Array [
            Object {
              "$$": Symbol(Atomico.$$),
              "children": Array [
                Array [
                  "This",
                ],
              ],
              "clone": undefined,
              "is": undefined,
              "key": undefined,
              "props": Object {},
              "raw": false,
              "shadow": undefined,
              "static": undefined,
              "type": "em",
            },
            " is ",
            Object {
              "$$": Symbol(Atomico.$$),
              "children": Array [
                Array [
                  "set test content",
                ],
              ],
              "clone": undefined,
              "is": undefined,
              "key": undefined,
              "props": Object {},
              "raw": false,
              "shadow": undefined,
              "static": undefined,
              "type": "strong",
            },
            " for ",
            Object {
              "$$": Symbol(Atomico.$$),
              "children": Array [
                Array [
                  Object {
                    "$$": Symbol(Atomico.$$),
                    "children": Array [
                      Array [
                        "the",
                      ],
                    ],
                    "clone": undefined,
                    "is": undefined,
                    "key": undefined,
                    "props": Object {},
                    "raw": false,
                    "shadow": undefined,
                    "static": undefined,
                    "type": "strong",
                  },
                ],
              ],
              "clone": undefined,
              "is": undefined,
              "key": undefined,
              "props": Object {},
              "raw": false,
              "shadow": undefined,
              "static": undefined,
              "type": "em",
            },
            " ",
            Object {
              "$$": Symbol(Atomico.$$),
              "children": Array [
                Array [
                  "remark",
                ],
              ],
              "clone": undefined,
              "is": undefined,
              "key": undefined,
              "props": Object {},
              "raw": false,
              "shadow": undefined,
              "static": undefined,
              "type": "code",
            },
            " hook for ",
            Object {
              "$$": Symbol(Atomico.$$),
              "children": Array [
                Array [
                  "atomico",
                ],
              ],
              "clone": undefined,
              "is": undefined,
              "key": undefined,
              "props": Object {},
              "raw": false,
              "shadow": undefined,
              "static": undefined,
              "type": "code",
            },
            ".",
          ],
        ],
        "clone": undefined,
        "is": undefined,
        "key": undefined,
        "props": Object {},
        "raw": false,
        "shadow": undefined,
        "static": undefined,
        "type": "p",
      },
    ],
  ],
  "clone": undefined,
  "is": undefined,
  "key": undefined,
  "props": Object {},
  "raw": false,
  "shadow": undefined,
  "static": undefined,
  "type": "div",
}
`;

exports[`Remark Hook for Atomico hook, tree from set content returns a result object when run 5`] = `
<html>
  <head />
  <body>
    <div>
      Mark {}
      <div>
        Mark {}
        <h1>
          Mark {}
          Set Test Content
          Mark {}
        </h1>
        

        <p>
          Mark {}
          <em>
            Mark {}
            This
            Mark {}
          </em>
           is 
          <strong>
            Mark {}
            set test content
            Mark {}
          </strong>
           for 
          <em>
            Mark {}
            <strong>
              Mark {}
              the
              Mark {}
            </strong>
            Mark {}
          </em>
           
          <code>
            Mark {}
            remark
            Mark {}
          </code>
           hook for 
          <code>
            Mark {}
            atomico
            Mark {}
          </code>
          .
          Mark {}
        </p>
        Mark {}
      </div>
      Mark {}
    </div>
  </body>
</html>
`;
